File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {

        int n=height.size();
        int left[n],right[n];
        left[0]=0;
        right[n-1]=0;

        for(int i=1;i<n;i++){
            left[i]=max(left[i-1],height[i-1]);
            
        }
        for(int i=n-2;i>=0;i--){
            right[i]=max(right[i+1],height[i+1]);

        }
        int water=0;int minheight=0;

        for(int i=0;i<n;i++){
            minheight=min(right[i],left[i]);
            if(minheight-height[i]>=0){

                 water+=(minheight-height[i]);
            }
           

        }
        return water;
        
    }
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:

        left=[]
        right=[]

        leftmax=float("-inf")
        for a in height:
            leftmax=max(leftmax,a)
            left.append(leftmax)
        rightmax=float("-inf")
        for a in reversed(height):
            rightmax=max(rightmax,a)
            right.append(rightmax)
        right=right[::-1]
        ans=0
        for i in range(0,len(height)):
            ans+=min(left[i],right[i])-height[i]
        return ans
        